### DISCUSSION TITLE
Enabling OpenMP under WSL

### DISCUSSION THREAD

**Discussion by walkersw** [firedrakeproject/firedrake#2616] [Firedrake support] [answered]
So I have Firedrake working under WSL, but it seems to only want OMP_NUM_THREADS = 1.  I ran lscpu, and it clearly knows I have more cores/threads.  Is there something else to do when installing Firedrake to enable more threads?

**Comment by rckirby** [comment, marked as answer]
If you want more cores, use mpiexec.  That is easier to support than a hybrid parallel model and actually helps with data locality

**Reply by walkersw** [reply]
Ok, I did that, and it does run in parallel, but it just runs the same code in parallel.  It also still says I have OMP_NUM_THREADS set greater than 1.  I assume I need to modify the Firedrake script to have a single simulation run, where the solving is done in parallel.

**Reply by connorjward** [reply]
I don't understand what you mean here. The way to use Firedrake in parallel is to write a single script as though it would be run in serial and then run it with the command `mpiexec -n <# processors> python mycode.py`. This will automatically run your code in parallel and do things like distributing the mesh between processors. There's more information [here](https://firedrakeproject.org/parallelism.html).

The reason you are seeing the warning about `OMP_NUM_THREADS` is because Firedrake only supports distributed memory parallelism, not shared memory parallelism (e.g. via OpenMP). In the past we've found that having `OMP_NUM_THREADS` greater than one can actually slow your code down which is why we emit this warning. The easiest way to make it go away is include the line `export OMP_NUM_THREADS=1` inside your `~/.bashrc` file.

**Reply by walkersw** [reply]
Ok, I was mistaken because I saw a lot of repeat print statements; I thought this meant it was running the same code multiple times (I also found the demo on printing in parallel).

But I see that the number of SNES iterations was not duplicated, so it is working as it should.
